app/src/main/java/com/infomaniak/mail/data/cache/MailboxContentController.kt

```diff
@@ -0,0 +1,147 @@
+/*
+ * Infomaniak kMail - Android
+ * Copyright (C) 2022 Infomaniak Network SA
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.infomaniak.mail.data.cache
+
+import com.infomaniak.mail.data.models.Folder
+import com.infomaniak.mail.data.models.message.Message
+import com.infomaniak.mail.data.models.thread.Thread
+import io.realm.MutableRealm
+import io.realm.MutableRealm.UpdatePolicy
+import io.realm.RealmResults
+import io.realm.query
+
+object MailboxContentController {
+
+    /**
+     * Folders
+     */
+    fun getFolders(): RealmResults<Folder> =
+        MailRealm.mailboxContent.query<Folder>().find()
+
+    fun getFolder(id: String): Folder? =
+        MailRealm.mailboxContent.query<Folder>("${Folder::id.name} == '$id'").first().find()
+
+    fun upsertFolder(folder: Folder): Folder =
+        MailRealm.mailboxContent.writeBlocking { copyToRealm(folder, UpdatePolicy.ALL) }
+
+    // fun updateFolder(id: String, onUpdate: (folder: Folder) -> Unit) {
+    //     MailRealm.mailboxContent.writeBlocking { getLatestFolder(id)?.let(onUpdate) }
+    // }
+
+    fun deleteFolder(id: String) {
+        MailRealm.mailboxContent.writeBlocking { getLatestFolder(id)?.let(::delete) }
+    }
+
+    private fun MutableRealm.getLatestFolder(id: String): Folder? =
+        getFolder(id)?.let(::findLatest)
+
+    /**
+     * Threads
+     */
+    fun getFolderThreads(folderId: String): List<Thread> =
+        MailRealm.mailboxContent.writeBlocking { getLatestFolder(folderId) }?.threads ?: emptyList()
+
+    fun getThread(uid: String): Thread? =
+        MailRealm.mailboxContent.query<Thread>("${Thread::uid.name} == '$uid'").first().find()
+
+    fun getLatestThread(uid: String): Thread? =
+        MailRealm.mailboxContent.writeBlocking { getLatestThread(uid) }
```

Comment:
infinite loop ?

Reply:
The 2nd one is an extension function of `MutableRealm`